cpp/src/sampling/random_walks_impl.cuh

```diff
@@ -255,147 +628,291 @@ random_walk_impl(raft::handle_t const& handle,
+                              std::ignore) =
+          cugraph::groupby_gpu_id_and_shuffle_values(
+            handle.get_comms(),
+            thrust::make_zip_iterator(current_vertices.begin(),
+                                            current_gpu.begin(),
+                                            current_position.begin()),
+            thrust::make_zip_iterator(current_vertices.end(),
+                                            current_gpu.end(),
+                                            current_position.end()),
+            [key_func =
+               cugraph::detail::compute_gpu_id_from_int_vertex_t<vertex_t>{
+                 {vertex_partition_range_lasts.begin(), vertex_partition_range_lasts.size()},
+                 major_comm_size,
+                 minor_comm_size}] __device__(auto val) { return key_func(thrust::get<0>(val)); },
+            handle.get_stream());
+      }
+    }
+
+    //  Sort for nbr_intersection, must sort all together
+    if (previous_vertices) {
+      if constexpr (multi_gpu){
+        thrust::sort(handle.get_thrust_policy(), 
+                     thrust::make_zip_iterator(current_vertices.begin(),
+                                               (*previous_vertices).begin(),
+                                               current_position.begin(),
+                                               current_gpu.begin()),
+                     thrust::make_zip_iterator(current_vertices.end(),
+                                               (*previous_vertices).end(),
+                                               current_position.end(),
+                                               current_gpu.end()));
+      } else {
+        thrust::sort(handle.get_thrust_policy(), 
+                     thrust::make_zip_iterator(current_vertices.begin(),
+                                               (*previous_vertices).begin(),
+                                               current_position.begin()),
+                     thrust::make_zip_iterator(current_vertices.end(),
+                                               (*previous_vertices).end(),
+                                               current_position.end()));
+      }
     }
 
     std::tie(current_vertices, new_weights) =
-      random_selector.follow_random_edge(handle, graph_view, edge_weight_view, current_vertices);
+      random_selector.follow_random_edge(handle, graph_view, edge_weight_view, current_vertices, previous_vertices);
 
     // FIXME: remove_if has a 32-bit overflow issue
     // (https://github.com/NVIDIA/thrust/issues/1302) Seems unlikely here (the goal of
     // sampling is to extract small graphs) so not going to work around this for now.
     CUGRAPH_EXPECTS(
       current_vertices.size() < static_cast<size_t>(std::numeric_limits<int32_t>::max()),
       "remove_if will fail, current_vertices.size() is too large");
-
+    size_t compacted_length{0};
```

Comment:
Scope of this seems off.  Make it as small as possible, perhaps inside the constexpr(multi_gpu) right below this.

Reply:
Compacted length is needed regardless of single or multi-GPU, it is used to remove invalid vertex IDs from the results span and resize the number of walks for future steps (starting Line 834).  The scope was expanded to cut down on code duplication, as the original code had the resize calls in each if-else branch block.

Reply:
OK.  I see that both paths through this if set it.